src/Datadog.Trace/AppSec/Waf/Waf.cs

```diff
@@ -66,53 +78,111 @@ public void Dispose(bool disposing)
+        private static JToken GetJsonReaderFromStream(Stream stream)
         {
-            try
+            var reader = new JsonTextReader(new StreamReader(stream));
+            return JToken.ReadFrom(reader);
+        }
+
+        private static Stream GetRulesManifestStream()
+        {
+            var assembly = typeof(Waf).Assembly;
+            return assembly.GetManifestResourceStream("Datadog.Trace.AppSec.Waf.rule-set.json");
+        }
+
+        private static Stream GetRulesFileStream(string rulesFile)
+        {
+            if (!File.Exists(rulesFile))
             {
-                DdwafConfigStruct args = default;
+                Log.Error($"AppSec could not find the rules file in path \"{rulesFile}\". AppSec will not run any protections in this application.");
+                return null;
+            }
 
-                var argCache = new List<Obj>();
-                var rules = RuleSetFromManifest(argCache);
+            return File.OpenRead(rulesFile);
+        }
 
-                var ruleHandle = WafNative.Init(rules.RawPtr, ref args);
+        private static WafHandle CreateWafHandle(string rulesFile)
+        {
+            var argCache = new List<Obj>();
+            Obj configObj;
+            try
+            {
+                var stream = GetRulesStream(rulesFile);
 
-                rules.Dispose();
-                foreach (var arg in argCache)
+                if (stream == null)
                 {
-                    arg.Dispose();
+                    return null;
                 }
 
-                if (ruleHandle == IntPtr.Zero)
+                if (Log.IsEnabled(LogEventLevel.Debug))
                 {
-                    Log.Error("Failed to create rules.");
-                    return null;
+                    LogRuleDetails(rulesFile);
                 }
-                else
+
+                configObj = CreatObjFromRulesStream(argCache, stream);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, $"AppSec could not read the rule file \"{rulesFile}\" as it was invalid. AppSec will not run any protections in this application.");
+                return null;
+            }
+
+            try
+            {
+                DdwafConfigStruct args = default;
+                var ruleHandle = WafNative.Init(configObj.RawPtr, ref args);
+                return new WafHandle(ruleHandle);
+            }
+            finally
+            {
+                configObj?.Dispose();
+                foreach (var arg in argCache)
                 {
-                    Log.Information("Rules successfully created.");
+                    arg.Dispose();
                 }
+            }
+        }
 
-                return new WafHandle(ruleHandle);
+        private static void LogRuleDetails(string rulesFile)
+        {
+            try
+            {
+                // because of the generic way rules are encoded for the WAF there's no good
```

Comment:
I think you can avoid the double parsing if you split `CreatObjFromRulesStream()` in two, pass the `JToken` into this method, and then call `Encoder.Encode` on the `JToken` subsequently in `CreateWafHandle`?

Reply:
Yes, you're right. I was more or less convinced JToken was forward only so double parsing was necessary, but it seems JToken is not forward only. Fixed in d4b9276affae22d7a61303e3c6cbf7718b515fad